pkg/controllers/targetversion/target_version_controller.go

```diff
@@ -0,0 +1,436 @@
+		return nil, nil
+	}
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{
+			{
+				Type:    "OAuthVersionRouteDegraded",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "GetFailed",
+				Message: fmt.Sprintf("Unable to get oauth-openshift route: %v", err),
+			},
+		}
+	}
+	return route, nil
+}
+
+func (c *targetVersionController) getIngressConfig() (*configv1.Ingress, []operatorv1.OperatorCondition) {
+	ingress, err := c.ingressLister.Get("cluster")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{{
+			Type:    "OAuthVersionIngressConfigDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NotFound",
+			Message: fmt.Sprintf("Unable to get cluster ingress config: %v", err),
+		}}
+	}
+	if len(ingress.Spec.Domain) == 0 {
+		return nil, []operatorv1.OperatorCondition{{
+			Type:    "OAuthVersionIngressConfigDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Invalid",
+			Message: fmt.Sprintf("The ingress config domain cannot be empty"),
+		}}
+	}
+	return ingress, nil
+}
+
+func (c *targetVersionController) getRouteSecret() (*corev1.Secret, []operatorv1.OperatorCondition) {
+	routerSecret, err := c.secretLister.Secrets("openshift-authentication").Get("v4-0-config-system-router-certs")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{{
+			Type:    "OAuthVersionRouteSecretDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "GetFailed",
+			Message: fmt.Sprintf("Unable to get OAuth server route certificate secret: %v", err),
+		}}
+	}
+	return routerSecret, nil
+}
+
+func (c *targetVersionController) getDeployment() (*appsv1.Deployment, []operatorv1.OperatorCondition) {
+	deployment, err := c.deploymentLister.Deployments("openshift-authentication").Get("oauth-openshift")
+	if err != nil {
+		return nil, []operatorv1.OperatorCondition{{
+			Type:    "OAuthVersionDeploymentDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "GetFailed",
+			Message: fmt.Sprintf("Unable to get OAuth server deployment: %v", err),
+		}}
+	}
+	return deployment, nil
+}
+
+func (c *targetVersionController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	foundConditions := []operatorv1.OperatorCondition{}
+
+	ingressConfig, ingressConditions := c.getIngressConfig()
+	foundConditions = append(foundConditions, ingressConditions...)
+
+	route, routeConditions := c.getRoute()
+	foundConditions = append(foundConditions, routeConditions...)
+
+	routeSecret, routeSecretConditions := c.getRouteSecret()
+	foundConditions = append(foundConditions, routeSecretConditions...)
+
+	deployment, deploymentConditions := c.getDeployment()
+	foundConditions = append(foundConditions, deploymentConditions...)
+
+	foundConditions = append(foundConditions, c.checkRouteHealthy(route, routeSecret, ingressConfig)...)
+	foundConditions = append(foundConditions, c.checkDeploymentReady(deployment)...)
+	foundConditions = append(foundConditions, c.oauthClientsReady(ctx)...)
```

Comment:
None of this belongs here, all of these checks should be spread to the controllers that have these resources as an output. 

Eventually, if you want to handle version like this, by a controller, you'll get the conditions from the state of the operator and set/not set the version based on that information.

Reply:
nvm, discard, I guess if we want to be checking these in parallel and not be stopped by other effects, this is fine

Reply:
i thought the purpose of this controller is to set version **only** is all these pre-checks are satisfied (iow. the operand is fully deployed and available). is that wrong assumption?

Reply:
it's a correct assumption, the source of those prereqs could be different (the operator's status itself?), but if a controller stops on not being able to retrieve route, then I guess it won't care about setting/unsetting the RouteHealthDegraded condition.

Reply:
status can be changed/cleared by external actors, so it is not the best thing to rely on (I did that in refactoring steps temporarily ;-)